src/vs/code/node/cli.ts

```diff
@@ -337,7 +362,61 @@ export async function main(argv: string[]): Promise<any> {
 			options['stdio'] = 'ignore';
 		}
 
-		const child = spawn(process.execPath, argv.slice(2), options);
+		let child: ChildProcess;
+		if (!isMacintosh) {
+			// We spawn process.execPath directly
+			child = spawn(process.execPath, argv.slice(2), options);
+		} else {
+			// On mac, we spawn using the open command to obtain behavior
+			// similar to if the app was launched from the dock
+			// https://github.com/microsoft/vscode/issues/102975
+
+			const spawnArgs = ['-n'];				// -n: launches even when opened already
+			spawnArgs.push('-a', process.execPath); // -a: opens a specific application
+
+			if (verbose) {
+				spawnArgs.push('--wait-apps'); // `open --wait-apps`: blocks until the launched app is closed (even if they were already running)
+
+				// The open command only allows for redirecting stderr and stdout to files,
+				// so we make it redirect those to temp files, and then use a logger to
+				// redirect the file output to the console
+				for (const outputType of ['stdout', 'stderr']) {
+
+					// Tmp file to target output to
+					const tmpName = createFileName(tmpdir(), `code-${outputType}`);
+					writeFileSync(tmpName, '');
+					spawnArgs.push(`--${outputType}`, tmpName);
+
+					// Listener to redirect content to stdout/stderr
+					processCallbacks.push(async (child: ChildProcess) => {
+						try {
+							const stream = outputType === 'stdout' ? process.stdout : process.stderr;
+
+							const cts = new CancellationTokenSource();
+							await Promise.race([
+								watchFileContents(tmpName, chunk => stream.write(chunk), cts.token),
+								Event.toPromise(Event.fromNodeEventEmitter(child, 'close')).then(() => cts.dispose(true))
```

Comment:
Now that we have a `CancellationToken` that is passed into the `watchFileContents` method, we no longer need to  use `Promise.race` actually. So I would suggest to change this code to `await watchFileContents...` and when the child fires the `close` event, call the `cts.dispose(true)` as we do now. That ensures that we are properly awaiting on the handle of `tmpName` getting closed in the `watchFileContents` method.